### PR TITLE
HD-1114: Support bind for kerberos/non-kerberos environment

### PR DESCRIPTION
**Pull request by sookim-altiscale**

### CODE REVIEW

File: metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java [hunk 1]
```diff
@@ -302,23 +302,14 @@ private void open() throws MetaException {
...
-      }
-
-      if (!hiveHost.isEmpty()) {
-        InetAddress localAddr = NetUtils.getLocalInetAddress(hiveHost);
-        if (localAddr != null) {
-          socket.bind(new InetSocketAddress(localAddr, 0));
-        }
+      UserGroupInformation loginUser = UserGroupInformation.getLoginUser();
+      if (loginUser != null && loginUser.hasKerberosCredentials()) {
+        hiveHost = SecurityUtil.getHostFromPrincipal(loginUser.getUserName());
```
Comment (raviprak-altiscale) [inline metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:312]: What happens in the case that we are not running Kerberos? hiveHost is never set. Is that acceptable?
I'll re-iterate. I am NOT familiar with this code. _Please please please post a patch in the open-source community_. I know its the longer route, but you _have_ to take it. Internally forking code has not worked well for most companies.

Comment (sookim-altiscale) [inline metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:312]: Yes. The original code is only have TSocket tsocket = new TSocket(store.getHost(), store.getPort(), 1000 \* clientSocketTimeout); I added it to bind the socket when it has kerberos principal. I will open the ticket on open-source community.

Comment (aw-was-here) [inline metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:312]: > Please please please post a patch in the open-source community.

I'm a bit annoyed this wasn't done already considering I specifically mentioned this had to be done last week.

Comment (sookim-altiscale) [inline metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:312]: That is my mistake. However, I was more focus on figuring out the one bug for non-kerberos environment to make working version. I will make a JIRA ticket today.

Comment (dbs-altiscale) [inline metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:312]: My question is why do you have to get the hostname from the principal?  Shouldn't it be correct in the configuration?
If it's different in the configuration, why can't we make it be the same?

File: metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java [hunk 2]
```diff
@@ -330,13 +321,14 @@ private void open() throws MetaException {
...
-            transport = new TSocket(socket);
-          } catch (TTransportException e) {
-              LOG.error("Couldn't create client transport to HiveMetastore(" 
-                + store.getHost() + ":" + store.getPort() + ")", e);
-              throw new MetaException(e.toString());
+            TSocket tsocket = new TSocket(store.getHost(), store.getPort(), 1000 * clientSocketTimeout);
+            if (hiveHost != null && !hiveHost.isEmpty())  {
+              socket = tsocket.getSocket();
+              InetAddress localAddr = NetUtils.getLocalInetAddress(hiveHost);
+              if (localAddr != null)
```
Comment (dbs-altiscale) [inline metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.java:328]: I still think this needs an 'else' clause to log the unexpected behavior of localAddr being null.  It's not going to work right in this case, will it?   If it does work right when localAddr is null, why do you need to do any of this?

**Pull request merged**